concurrency/ItemExpiryPurgerTaskMetricsCollector.java

```diff
@@ -0,0 +1,193 @@
+        addToSingleCountMetricMapEntry(metricName, 1 /* value */);
+    }
+
+    private synchronized void addToSingleCountMetricMapEntry(final String metricName, final Integer value) {
+        Integer metricValue = purgerCountMetricsMap.get(metricName);
+        bbAssert(metricValue != null, "addToSingleMetricMapEntry", "Metric not available", "metricName", metricName);
+        purgerCountMetricsMap.put(metricName, metricValue + value);
+    }
+    private synchronized void addToSingleTimeMetricMapEntry (final String metricName, final long value) {
+        ItemExpiryPurgerTimeMetricPair metricValue = purgerTimeMetricsMap.get(metricName);
+        bbAssert(metricValue != null, "addToSingleTimeMetricMapEntry", "Metric not available", "metricName", metricName);
+        metricValue.addItemExpiryPurgerStaleness(value);
+        purgerTimeMetricsMap.put(metricName, metricValue);
+    }
+
+    public void updateItemExpiryPurgerStaleness(final long staleness) {
+        addToSingleTimeMetricMapEntry(BBCoralMetrics.Field.ItemExpiryPurgerStaleness.name(), staleness);
+    }
+
+    public void increasePurgerTaskGetZeroNtpTime() {
+        incrementSingleCountMetricMapEntry(BBCoralMetrics.Field.ItemExpiryPurgerExpressionZero.name());
+    }
+
+    public void increasePurgerTaskFailedByException() {
+        incrementSingleCountMetricMapEntry(BBCoralMetrics.Field.ItemExpiryPurgerTaskFailedByException.name());
+    }
+
+    public void increasePurgerTaskThrottledByBaseTableIops() {
+        incrementSingleCountMetricMapEntry(BBCoralMetrics.Field.ItemExpiryPurgerTaskThrottledByBaseTableIops.name());
+    }
+
+    public void increasePurgerTaskThrottledByGsi() {
+        incrementSingleCountMetricMapEntry(BBCoralMetrics.Field.ItemExpiryPurgerTaskThrottledByGsi.name());
+    }
+
+    public void increasePurgerTaskThrottledByUpdateStreams() {
+        incrementSingleCountMetricMapEntry(BBCoralMetrics.Field.ItemExpiryPurgerTaskThrottledByUpdateStreams.name());
+    }
+
+    public void increasePurgerTaskThrottledByReplicaTableEntryThrottle() {
+        incrementSingleCountMetricMapEntry(BBCoralMetrics.Field.ItemExpiryPurgerTaskThrottledByReplicaTableEntryThrottle.name());
+    }
+
+    public void increasePurgerTaskConditionCheckFailed() {
+        incrementSingleCountMetricMapEntry(BBCoralMetrics.Field.ItemExpiryPurgerTaskConditionCheckFailed.name());
+    }
+
+    public void increasePurgerTaskUnexpiredItemDequeued() {
+        incrementSingleCountMetricMapEntry(BBCoralMetrics.Field.ItemExpiryPurgerTaskUnexpiredItemDequeued.name());
+    }
+
+    public void increaseItemExpiryWriteIOPsPublishedToStreams(int logicalIOCount) {
+        addToSingleCountMetricMapEntry(BBCoralMetrics.Field.ItemExpiryWriteIOPsPublishedToStreams.name(), logicalIOCount /* value */);
+    }
+
+    private void collectCountAndTimeMetrics(final OpContext opContext) {
+        for (Map.Entry<String, Integer> entry : purgerCountMetricsMap.entrySet()) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.



**Problem**  
 This line may not be thread-safe: **148**. At this line the following field is not protected by synchronization: `purgerCountMetricsMap`. Your code typically protects this field with synchronization on object: `ItemExpiryPurgerTaskMetricsCollector.this`.

**Fix**  
 Consider protecting the above field with synchronization on object: `ItemExpiryPurgerTaskMetricsCollector.this`. For example, like at lines: **161**, **173**, **169**, **177**, **165**, **96**, and **98**.

**More info**  
 The following lines are where the above field is already protected by synchronization on the above object: **47**, **48**, **49**, **50**, **51**, **52**, **53**, **54**, **55**, **56**, **57**, **58**, **96**, **98**, **161**, **165**, **169**, **173**, and **177**.